pkg/dyninst/decode/decoder.go

```diff
@@ -60,478 +73,119 @@ type typeAndAddr struct {
 	addr   uint64
 }
 
-// Decode decodes the given event into the given writer.
-func (d *Decoder) Decode(event output.Event, out io.Writer) error {
-	enc := jsontext.NewEncoder(out)
-	err := enc.WriteToken(jsontext.BeginObject)
+// Decode decodes the output Event from the BPF program into a JSON format to the specified output writer.
+func (d *Decoder) Decode(event Event, out io.Writer) error {
+	err := d.snapshotMessage.init(d, event)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not decode event: %w", err)
 	}
+	defer d.snapshotMessage.clear()
+	return json.MarshalWrite(out, &d.snapshotMessage)
+}
 
-	header, err := event.Header()
-	if err != nil {
-		return err
-	}
+// Event wraps the output Event from the BPF program. It also adds fields
+// that are not present in the BPF program.
+type Event struct {
+	output.Event
+	ServiceName string
+}
 
-	// bpf will only upload the pc's for a given hash once.
-	// We cache the frames for each hash accordingly.
-	var frames []uint64
-	frames, ok := d.stackHashes[header.Stack_hash]
-	if !ok {
-		frames, err = event.StackPCs()
-		if err != nil {
-			return err
-		}
-		d.stackHashes[header.Stack_hash] = frames
-	}
+type snapshotMessage struct {
+	Service  string       `json:"service"`
+	DDSource string       `json:"ddsource"`
+	DDTags   string       `json:"ddtags"`
+	Logger   logger       `json:"logger"`
+	Debugger debuggerData `json:"debugger"`
 
-	err = enc.WriteToken(jsontext.String("stack_frames"))
-	if err != nil {
-		return err
-	}
-	err = enc.WriteToken(jsontext.BeginArray)
-	if err != nil {
-		return err
-	}
-	for i := range frames {
-		err = enc.WriteToken(jsontext.Uint(frames[i]))
-		if err != nil {
-			return err
-		}
-	}
-	err = enc.WriteToken(jsontext.EndArray)
-	if err != nil {
-		return err
-	}
+	addressReferenceCount map[typeAndAddr]output.DataItem
+	rootData              []byte
+}
 
-	err = enc.WriteToken(jsontext.String("captures"))
-	if err != nil {
-		return err
+func (s *snapshotMessage) init(decoder *Decoder, event Event) error {
+	s.Service = event.ServiceName
+	s.Debugger.Snapshot = snapshotData{
+		decoder:  decoder,
+		ID:       uuid.New(),
+		Language: "go",
 	}
-	err = enc.WriteToken(jsontext.BeginObject)
-	if err != nil {
-		return err
-	}
-	addressReferenceCount := map[typeAndAddr]output.DataItem{}
 
-	var (
-		rootData []byte
-		rootType *ir.EventRootType
-	)
-	for item, err := range event.DataItems() {
+	var rootType *ir.EventRootType
+	s.addressReferenceCount = map[typeAndAddr]output.DataItem{}
```

Comment:
rather than constructing a new map every time, how about we keep the old one around so we don't have to keep re-allocating the backing arrays?